### ISSUE TITLE
Folder hierarchy in BSP2 "Bookmarks Menu" drop down wildly different than FFx library

### ISSUE THREAD

**Issue by ttggzz** [aaFn/Bookmark-search-plus-2#59]
Hi, is the folder hierarchy in bsp2 supposed to match the FFx library?

Using FFx 62.0.3 on Linux, everything up to date. I just went to do a search using BSP2 and noticed the drop down list of folders is radically different than the way I have them organized in the actual firefox library. I had not used bsp2 for 6 weeks or so, got back to it a few days ago and I believe the bsp2 folders at that time matched the library [I could be wrong there, but I thought I had tested dragging stuff around from the library into bsp2].

Past few days I had done a huge amount of bookmarking and reorganizing in the library, no use of bsp2. Today when I went to do a search the bsp2 folders are wonked out. I restarted firefox and the computer, no change.
thanks

**Comment by aaFn** [comment]
Hello @ttggzz , could you clarify what you mean by "no use of BSP2" ?
Do you mean you deactivated it ?

If yes this is normal ... due to a number of limitations in FF, in particular a very slow API to get bookmarks
cf. https://bugzilla.mozilla.org/show_bug.cgi?id=1396364
I had to rely solely on notification events of modifications after an initial load (long if plenty of bookmarks) from the FF API, and not using the FF API anymore across restarts, but storing them in local storage to retrieve them very quickly.

Which means that if you deactivate BSP2, it doesn't get the notifications of modifications, and then it is out of sync, and so the hierarchy or contents do not match.

The remediation is to go to the Options page for BSP2, and to click on the "Reload from FF API now" button .. which can be very long if you have lots of bookmarks (> 10000).

I didn't find a way to know / be notified that the add-on was just reactivated after a deactivate, to trigger an automatic reload, so it must be manual so far, sorry :-(

Let me know.

**Comment by ttggzz** [comment]
ahh, OK thanks, I had actually looked at that "reload from api" thingy before posting and wondered about it... Regarding my "no use of bsp2", I'm not sure if I disabled it. I believe I may have, because there has been no lagging on my machine until today [if you see the other issue I posted today "bsp2 uses resources when it's not open"].

OK, I will proceed with this info and work with it manually. Thanks for your work, it is mind boggling that Mozilla does not have this as a native feature in Firefox. Astonishing.
